packages/discovery-provider/src/queries/get_extended_purchase_gate.py

```diff
@@ -0,0 +1,240 @@
+import logging
+from datetime import datetime
+from typing import Dict, List, Optional, TypedDict, Union, cast
+
+from sqlalchemy import and_, func
+from sqlalchemy.orm import aliased
+from sqlalchemy.orm.session import Session
+
+from src.models.playlists.playlist import Playlist
+from src.models.tracks.track import Track
+from src.models.users.user import User
+from src.models.users.user_bank import USDCUserBankAccount
+from src.models.users.user_payout_wallet_history import UserPayoutWalletHistory
+from src.utils.db_session import get_db_read_replica
+from src.utils.session_manager import SessionManager
+
+logger = logging.getLogger(__name__)
+
+
+class GetPurchaseInfoArgs(TypedDict):
+    content_id: int
+    content_type: str
+    content: Union[Track, Playlist]
+
+
+class Split(TypedDict):
+    user_id: int
+    percentage: float
+
+
+class USDCPurchaseCondition(TypedDict):
+    price: int
+    splits: List[Split]
+
+
+class PurchaseGate(TypedDict):
+    usdc_purchase: USDCPurchaseCondition
+
+
+class ExtendedSplit(Split):
+    amount: int
+    payout_wallet: str
+    eth_wallet: str
+
+
+class ExtendedUSDCPurchaseCondition(TypedDict):
+    price: int
+    splits: List[ExtendedSplit]
+
+
+class ExtendedPurchaseGate(TypedDict):
+    usdc_purchase: ExtendedUSDCPurchaseCondition
+
+
+class LegacyUSDCPurchaseCondition(TypedDict):
+    price: int
+    splits: Dict[str, int]
+
+
+class LegacyPurchaseGate(TypedDict):
+    usdc_purchase: LegacyUSDCPurchaseCondition
+
+
+class FollowGate(TypedDict):
+    follow_user_id: int
+
+
+class TipGate(TypedDict):
+    tip_user_id: int
+
+
+class NFTCollection(TypedDict):
+    chain: str
+    address: str
+    name: str
+    imageUrl: Optional[str]
+    externalLink: Optional[str]
+
+
+class NFTGate(TypedDict):
+    nft_collection: NFTCollection
+
+
+AccessGate = Union[PurchaseGate, FollowGate, TipGate, NFTGate]
+
+
+# Allow up to 6 decimal points for split percentages
+percentage_decimals = 6
+percentage_multiplier = 10**percentage_decimals
+
+
+cents_to_usdc_multiplier = 10**4
+
+
+def calculate_split_amounts(price: int, splits: List[Split]):
+    """
+    Deterministically calculates the USDC amounts to pay to each person,
+    adjusting for rounding errors and ensuring the total matches the price.
+    """
+    price_in_usdc = int(cents_to_usdc_multiplier * price)
+    running_total = 0
+    new_splits: List[Dict] = []
+    for index in range(len(splits)):
+        split = splits[index]
+        # multiply percentage to make it a whole number
+        percentage_whole = int(split["percentage"] * percentage_multiplier)
+        # do safe integer math on the price
+        amount = int(percentage_whole * price_in_usdc)
```

Comment:
why is it necessary to cast this to an int here? it can't be a decimal right? given that it's just cents * percentage whole. can we do an assert amount == int(amount)?

Reply:
true, I think this was just FUD trying to avoid floats as much as possible